### ISSUE TITLE
else if without conditional fails to compile

### ISSUE THREAD

**Issue by kennethlimcp** [spark/buildpack-wiring-preprocessor#6]
From: https://github.com/spark/docs/pull

For some reason this fails:

```
boolean var1 = false;
void loop() {
 if(var1) {
 }
 else if(var1) {
 }
}
```

and this works:

```
boolean var1 = false;
void loop() {
 if(var1) {
 }
 else if(var1 == true) {
 }
}
```

**Comment by pwallington** [comment]
This fails because the RegExp at `cparser/regexParser.js:60` for matching function definitions is too permissive (see expanded form below).  Parsing C is hard and DFA regexps can't do it completely.
I have a couple of ideas for how to prevent this - but I don't know what the load impact on your build farm would be:
  * Pre-parse the string to remove function blocks, so only definitions remain to be matched.
  * Check matched function name definitions against C reserved words to exclude matches like `else if(foo) {`
(These could also be combined as they work independently)

```
		definitions: function definitions(str) {
			var fnRegex = new RegExp("[\\w\\[\\]\\*]+\\s+[&\\[\\]\\*\\w\\s]+\\([&,\\[\\]\\*\\w\\s]*\\)(?=\\s*\\{)", 'gm');
			return that.matchAll(fnRegex, str);
```
RegExp expanded with extra `\`s removed:
`[\w\[\]\*]+` - one or more word characters, plus `[`, `]`, `*`
`\s+` - one or more whitespace
`[&\[\]\*\w\s]+` - one or more whitespace or word characters, plus `&`, `[`, `]`, `*`
`\([&,\[\]\*\w\s]*\)` - zero or more of the same chars, inside parens
`(?=\s*\{)` - followed by (not included in match) zero or more whitespace and `{` (start of function block)

Another potential issue with this regexp (and others like it in the file) is multiple one-or-more whitespace matches can result in exponentially bad backtracking performance for failing matches [see this StackOverflow postmortem as a reference](http://stackstatus.net/post/147710624694/outage-postmortem-july-20-2016).

Finally - double-escaping `\`s can be eliminated by putting regexps inside `/`s instead of double-quotes.

**Comment by monkbroc** [comment]
Fixed in #7 

Thanks for the insightful comment @pwallington. I went the easy route of removing the special case `else if(foo)` rather than improve the original regexp which as you say can't parse C perfectly anyways.

**Comment by monkbroc** [comment]
Just to be clear, this is fixed on the prerelease firmware 0.6.1-rc.1 and later only and will be rolled out shortly to the other firmware versions.
